README.md
# cloudwatch-logs-actions

[![Build Status](https://github.com/mooyoul/cloudwatch-logs-actions/workflows/workflow/badge.svg)](https://github.com/mooyoul/cloudwatch-logs-actions/actions)
[![Example Status](https://github.com/mooyoul/cloudwatch-logs-actions/workflows/example/badge.svg)](https://github.com/mooyoul/cloudwatch-logs-actions/actions)
[![Semantic Release enabled](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com/)
[![MIT license](http://img.shields.io/badge/license-MIT-blue.svg)](http://mooyoul.mit-license.org/)

GitHub action that integrates with Amazon CloudWatch Logs.

![Example](./assets/example.png)

-----

## Why?

I have an open sourced Backend Project, and I want to hide
deploy log to prevent sensitive information leak.

... but currently, There are no way to hide logs of Github Actions.
So, My idea is - forward stdout/stderr output to CloudWatch Logs without printing logs.

See below example for further details.


##### Example

```yaml
# ...
jobs:
  job:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Deploy
        uses: mooyoul/cloudwatch-logs-actions@v1.1.5
        env:
          AWS_DEFAULT_REGION: us-east-1
          AWS_REGION: us-east-1
          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
        with:
          group: my-open-project
          stream: ${{ gihub.sha }}
          retention: 14
          run: |
            # you can use multiline commands if needed
            npm run deploy:stage
            npm run deploy:prod
```

##### Input

```typescript
type Input = {
  // Cloudwatch Logs Region
  region: string;

  // Name of Log Group
  group: string;

  // Name of Log Stream
  stream: string;

  // The number of days to retain the log events in the specified log group.
  // Possible values are: 1, 3, 5, 7, 14, 30, 60, 90, 120, 150, 180, 365, 400, 545, 731, 1827, and 3653.
  retention?: string;
  
  // Which shell to use
  // Possible values are: sh, bash (default: sh)
  shell?: string;

  // Command to run, Supports multiple commands
  run: string;
}
```

##### Output

None

## License

[MIT](LICENSE)

See full license on [mooyoul.mit-license.org](http://mooyoul.mit-license.org/)


CHANGELOG.md
## [1.1.5](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.1.4...v1.1.5) (2020-06-26)


### Bug Fixes

* **consumer:** respect returned nextSequenceToken value ([5b395db](https://github.com/mooyoul/cloudwatch-logs-actions/commit/5b395db4ad1f0f5344119f2ea5d4036ef238c7ce))

## [1.1.4](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.1.3...v1.1.4) (2020-06-26)


### Bug Fixes

* **consumer:** ignore empty line text ([4c514ac](https://github.com/mooyoul/cloudwatch-logs-actions/commit/4c514ac23994be6f56dfc4f22433addbb4fc4f10))

## [1.1.3](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.1.2...v1.1.3) (2020-06-26)


### Bug Fixes

* **runner:** fix flush cannot be called if exec throws error ([4e87cb2](https://github.com/mooyoul/cloudwatch-logs-actions/commit/4e87cb2c11e4a860635cf5ced80af70fb8b8abfb))

## [1.1.2](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.1.1...v1.1.2) (2020-06-26)


### Bug Fixes

* **flush:** fix last flush can be ignored ([f9907be](https://github.com/mooyoul/cloudwatch-logs-actions/commit/f9907be4d10c557107f451b7ce1c0aa86b34f527))

## [1.1.1](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.1.0...v1.1.1) (2020-06-26)


### Bug Fixes

* **flush:** fix empty log events flush attempts ([e4d3d73](https://github.com/mooyoul/cloudwatch-logs-actions/commit/e4d3d737223e5e9d257bf99c778ffc87d13203e2))

# [1.1.0](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.0.1...v1.1.0) (2020-06-26)


### Bug Fixes

* **consumer:** fix shouldFlush returns true even if buffer is empty ([1323e87](https://github.com/mooyoul/cloudwatch-logs-actions/commit/1323e8773651ef4ee85f0634c7c434b00f448ccb))


### Features

* **retention:** add log retention support ([14f9cb6](https://github.com/mooyoul/cloudwatch-logs-actions/commit/14f9cb60c5f045a6e5d9d8f10abe405fba35aeaa))

## [1.0.1](https://github.com/mooyoul/cloudwatch-logs-actions/compare/v1.0.0...v1.0.1) (2020-06-24)


### Bug Fixes

* use shell based exec to support multiline run statements ([b858fb2](https://github.com/mooyoul/cloudwatch-logs-actions/commit/b858fb2780032a167d01f3bdd8c23281bcce5fca))

# 1.0.0 (2020-06-24)


### Features

* initial commit ([4f5668f](https://github.com/mooyoul/cloudwatch-logs-actions/commit/4f5668f8f98708631863c7d5c0b8aefab4e995dc))


test/index.test.ts
import * as sinon from "sinon";
import { sandbox, stubAWSAPI } from "./helper";

import * as CloudWatchLogs from "aws-sdk/clients/cloudwatchlogs";

import { CloudWatchLogsConsumer } from "../src/consumer";

describe("cloudwatch-logs-actions", () => {
  it("should work", async () => {
    const fakeTimer = sandbox.useFakeTimers(1000);

    const createLogGroupFake = sinon.fake.resolves({});
    stubAWSAPI(CloudWatchLogs, "createLogGroup", createLogGroupFake);

    const putRetentionPolicyFake = sinon.fake.resolves({});
    stubAWSAPI(CloudWatchLogs, "putRetentionPolicy", putRetentionPolicyFake);

    const createLogStreamFake = sinon.fake.resolves({});
    stubAWSAPI(CloudWatchLogs, "createLogStream", createLogStreamFake);

    const putLogEventsFake = sinon.fake.resolves({ nextSequenceToken: "token" });
    stubAWSAPI(CloudWatchLogs, "putLogEvents", putLogEventsFake);

    const res = new CloudWatchLogsConsumer({
      region: "us-east-1",
      group: "group-name",
      stream: "stream-name",
      retentionInDays: 7,
    });

    await res.consume("foo");
    expect(createLogGroupFake.callCount).toEqual(0);
    expect(putRetentionPolicyFake.callCount).toEqual(0);
    expect(createLogStreamFake.callCount).toEqual(0);
    expect(putLogEventsFake.callCount).toEqual(0);

    fakeTimer.tick(60000);

    await res.flush();
    expect(createLogGroupFake.callCount).toEqual(1);
    expect(createLogGroupFake.firstCall.args).toEqual([{
      logGroupName: "group-name",
    }]);

    expect(putRetentionPolicyFake.callCount).toEqual(1);
    expect(putRetentionPolicyFake.firstCall.args).toEqual([{
      logGroupName: "group-name",
      retentionInDays: 7,
    }]);

    expect(createLogStreamFake.callCount).toEqual(1);
    expect(createLogStreamFake.firstCall.args).toEqual([{
      logGroupName: "group-name",
      logStreamName: "stream-name",
    }]);
    expect(putLogEventsFake.callCount).toEqual(1);
    expect(putLogEventsFake.firstCall.args).toEqual([{
      logGroupName: "group-name",
      logStreamName: "stream-name",
      logEvents: [{
        message: "foo",
        timestamp: 1000,
      }],
    }]);

    await res.flush();
    await res.flush();
    await res.flush();

    expect(createLogGroupFake.callCount).toEqual(1);
    expect(putRetentionPolicyFake.callCount).toEqual(1);
    expect(createLogStreamFake.callCount).toEqual(1);
    expect(putLogEventsFake.callCount).toEqual(1);

    await res.consume(""); // empty line
    await res.flush();

    expect(createLogGroupFake.callCount).toEqual(1);
    expect(putRetentionPolicyFake.callCount).toEqual(1);
    expect(createLogStreamFake.callCount).toEqual(1);
    expect(putLogEventsFake.callCount).toEqual(1);

    await res.consume("bar");

    fakeTimer.tick(60000);

    await res.consume("baz");

    await res.flush();

    expect(createLogGroupFake.callCount).toEqual(1);
    expect(putRetentionPolicyFake.callCount).toEqual(1);
    expect(createLogStreamFake.callCount).toEqual(1);

    expect(putLogEventsFake.callCount).toEqual(2);
    expect(putLogEventsFake.secondCall.args).toEqual([{
      logGroupName: "group-name",
      logStreamName: "stream-name",
      logEvents: [{
        message: "bar",
        timestamp: 61000,
      }, {
        message: "baz",
        timestamp: 121000,
      }],
      sequenceToken: "token",
    }]);
  });
});


src/consumer.ts
import { Sema } from "async-sema";
import * as CloudWatchLogs from "aws-sdk/clients/cloudwatchlogs";

export type CloudWatchLogsConsumerOptions = {
  region: string;
  group: string;
  stream: string;
  retentionInDays?: number;
};

const MAX_RECORDS = 5000;
const MAX_SIZE = 1024 * 800; // 800 KB
const MAX_DELAY = 5000; // 5 sec

export class CloudWatchLogsConsumer {
  private readonly cwlogs: CloudWatchLogs;
  private readonly group: string;
  private readonly stream: string;
  private readonly retentionInDays?: number;

  private readonly sema = new Sema(1, { capacity: 512 });
  private flushedAt: number = Date.now();
  private initialized = false;
  private sequenceToken?: string;

  private buffer: CloudWatchLogs.InputLogEvent[] = [];
  private bufferSize: number = 0;

  public constructor(options: CloudWatchLogsConsumerOptions) {
    this.cwlogs = new CloudWatchLogs({ region: options.region });
    this.group = options.group;
    this.stream = options.stream;
    this.retentionInDays = options.retentionInDays;
  }

  public async consume(line: string) {
    if (line.length > 0) {
      this.queue(line);

      if (this.shouldFlush) {
        await this.flush();
      }
    }
  }

  public async flush() {
    if (this.flushable) {
      const events = this.buffer;
      this.buffer = [];
      this.bufferSize = 0;

      await this.sema.acquire();

      try {
        await this.prepareStream();
        const res = await this.cwlogs.putLogEvents({
          logGroupName: this.group,
          logStreamName: this.stream,
          logEvents: events,
          sequenceToken: this.sequenceToken,
        }).promise();

        this.flushedAt = Date.now();
        this.sequenceToken = res.nextSequenceToken;
      } finally {
        this.sema.release();
      }
    }
  }

  private async prepareStream() {
    if (!this.initialized) {
      this.initialized = true;

      await this.cwlogs
        .createLogGroup({
          logGroupName: this.group,
        })
        .promise()
        .catch((e) => e.name === "ResourceAlreadyExistsException" ? Promise.resolve() : Promise.reject(e));

      if (this.retentionInDays !== undefined) {
        await this.cwlogs
          .putRetentionPolicy({
            logGroupName: this.group,
            retentionInDays: this.retentionInDays,
          })
          .promise();
      }

      await this.cwlogs
        .createLogStream({
          logGroupName: this.group,
          logStreamName: this.stream,
        })
        .promise()
        .catch((e) => e.name === "ResourceAlreadyExistsException" ? Promise.resolve() : Promise.reject(e));
    }
  }

  private get flushable() {
    return this.buffer.length > 0;
  }

  private get shouldFlush() {
    if (this.buffer.length > MAX_RECORDS) {
      return true;
    }

    if (this.bufferSize > MAX_SIZE) {
      return true;
    }

    return Date.now() - this.flushedAt > MAX_DELAY;
  }

  private queue(line: string) {
    const size = Buffer.byteLength(line) + 26;

    this.buffer.push({
      message: line,
      timestamp: Date.now(),
    });
    this.bufferSize += size;
  }
}


src/index.ts
import * as core from "@actions/core";
import { exec } from "@actions/exec";

import { CloudWatchLogsConsumer } from "./consumer";

const ALLOWED_RETENTION_DAYS = new Set<number>([
  1, 3, 5, 7, 14, 30, 60, 90, 120, 150, 180, 365, 400, 545, 731, 1827, 3653,
]);

type Shell = {
  program: string;
  args: string[];
};

(async () => {
  const region = core.getInput("region") || process.env.AWS_REGION || "us-east-1";
  const group = core.getInput("group");
  const stream = core.getInput("stream");
  const retentionInDays = (() => {
    const raw = core.getInput("retention");
    const parsed = parseInt(raw, 10);

    return ALLOWED_RETENTION_DAYS.has(parsed) ? parsed : undefined;
  })();
  const run = core.getInput("run");
  const shellName = core.getInput("shell") || "sh";

  const shell = getShell(shellName);

  const consumer = new CloudWatchLogsConsumer({
    region,
    group,
    stream,
    retentionInDays,
  });

  try {
    await exec(shell.program, shell.args, {
      input: Buffer.from(run, "utf8"),
      silent: true,
      listeners: {
        stdline(line: string) {
          consumer.consume(line).catch(() => { /* swallow error */ });
        },
        errline(line: string) {
          consumer.consume(line).catch(() => { /* swallow error */ });
        },
      },
    });
  } finally {
    await consumer.flush()
      .catch((e) => {
        // tslint:disable-next-line
        console.error("Failed to flush: ", e.stack);
      });
  }
})().catch((e) => {
  console.error(e.stack); // tslint:disable-line
  core.setFailed(e.message);
});


function getShell(name: string): Shell {
  const normalizedName = name.toLowerCase();

  switch (normalizedName) {
    case "bash":
      return {
        program: "bash",
        args: ["--noprofile", "--norc", "-eo", "pipefail"],
      };
    default:
      return {
        program: "sh",
        args: ["-e"],
      };
  }
}
